File: src/pcl-bindings/GaussianKernelPointXYZPointXYZ.cpp
#include "export.h"
#include "pcl/pcl_base.h"
#include "pcl/point_types.h"
#include <pcl/filters/convolution_3d.h>

using namespace pcl;
using namespace std;
using namespace Eigen;
using namespace pcl::filters;

using point_t = PointXYZ;
using kernel_t = GaussianKernel<point_t, point_t>;

EXPORT(kernel_t*) gaussian_kernel_pointxyz_pointxyz_ctor()
{
    return new kernel_t();
}

EXPORT(void) gaussian_kernel_pointxyz_pointxyz_delete(kernel_t** ptr)
{
    delete* ptr;
    *ptr = NULL;
}

EXPORT(void) gaussian_kernel_pointxyz_pointxyz_set_sigma(kernel_t* ptr, float sigma)
{
    ptr->setSigma(sigma);
}

EXPORT(void) gaussian_kernel_pointxyz_pointxyz_set_threshold_relative_to_sigma(kernel_t* ptr, float sigma_coefficient)
{
    ptr->setThresholdRelativeToSigma(sigma_coefficient);
}

EXPORT(bool) gaussian_kernel_pointxyz_pointxyz_init_compute(kernel_t* ptr)
{
    return ptr->initCompute();
}

EXPORT(void) gaussian_kernel_pointxyz_pointxyz_set_input_cloud(kernel_t* ptr, PointCloud<point_t>* cloud)
{
    PointCloud<point_t>::Ptr shared(std::make_shared<PointCloud<point_t>>(*cloud));
    ptr->setInputCloud(shared);
}

EXPORT(void) gaussian_kernel_pointxyz_pointxyz_set_threshold(kernel_t* ptr, float threshold)
{
    ptr->setThreshold(threshold);
}

File: src/Pcl.NET/GaussianKernelPointXYZPointXYZ.cs
namespace Pcl.NET
{
    public class GaussianKernelPointXYZPointXYZ : GaussianKernel<PointXYZ, PointXYZ>
    {
        private float _sigma;
        private float _threshold;
        private float _thresholdRelativeToSigma;
        private PointCloud<PointXYZ>? _input;

        public override float Sigma
        {
            get
            {
                ThrowIfDisposed();
                return _sigma;
            }
            set
            {
                ThrowIfDisposed();
                Invoke.gaussian_kernel_pointxyz_pointxyz_set_sigma(_ptr, value);
                _sigma = value;
            }
        }
        public override float Threshold
        {
            get
            {
                ThrowIfDisposed();
                return _threshold;
            }
            set
            {
                ThrowIfDisposed();
                Invoke.gaussian_kernel_pointxyz_pointxyz_set_threshold(_ptr, value);
                _threshold = value;
            }
        }
        public override float ThresholdRelativeToSigma
        {
            get
            {
                ThrowIfDisposed();
                return _thresholdRelativeToSigma;
            }
            set
            {
                ThrowIfDisposed();
                Invoke.gaussian_kernel_pointxyz_pointxyz_set_threshold_relative_to_sigma(_ptr, value);
                _thresholdRelativeToSigma = value;
            }
        }
        public override PointCloud<PointXYZ>? Input
        {
            get
            {
                ThrowIfDisposed();
                return _input;
            }
            set
            {
                ThrowIfDisposed();
                Invoke.gaussian_kernel_pointxyz_pointxyz_set_input_cloud(_ptr, value);
                _input = value;
            }
        }

        public GaussianKernelPointXYZPointXYZ()
        {
            _ptr = Invoke.gaussian_kernel_pointxyz_pointxyz_ctor();
        }
        public override void InitCompute()
        {
            Invoke.gaussian_kernel_pointxyz_pointxyz_init_compute(_ptr);
        }

        protected override void DisposeObject()
        {
            if (!_suppressDispose)
            {
                Invoke.gaussian_kernel_pointxyz_pointxyz_delete(ref _ptr);
                _ptr = IntPtr.Zero;
            }
        }
    }
}
